File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxDepth(TreeNode node) {
        int ans = 0;
        if (node == null) {
            return 0;
        }
        int left = maxDepth(node.left);
        int right = maxDepth(node.right);
        ans = 1 + Math.max(left, right);
        return ans;
    }
}

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        /*Yeah so like draw the tree and track it,
        you would observe that if the depth is 2, leetcode shows 3,
        well teh actual depth is indeed 3, so basically our final answer would
        have a +1 while we return it, in fact that will be necesaary at every
        setp*/

        // Self-Explanatory
        if (root == nullptr) {
            return 0;
        }

        // We first traverse left then right and store the depth in two
        // variables
        int leftDepth = maxDepth(root->left);
        int rightDepth = maxDepth(root->right);

        // Now we return the max amongst the two with a +1
        return max(leftDepth, rightDepth) + 1;
    }
};